src/mod/pbkdf2.mod/pbkdf2.c

```diff
@@ -0,0 +1,302 @@
+  restlen = outlen;
+  bufcount(&out2, &restlen, snprintf((char *) out2, restlen,
+           "$pbkdf2-%s$rounds=%u$", digest_name, rounds));
+  ret = b64_ntop_without_padding(salt, saltlen, out2, restlen);
+  if (ret < 0) {
+    explicit_bzero(out, outlen);
+    putlog(LOG_MISC, "*", "PBKDF2 error: b64_ntop(salt).");
+    return NULL;
+  }
+  bufcount(&out2, &restlen, ret);
+  out2[0] = '$';
+  bufcount(&out2, &restlen, 1);
+  buf = nmalloc(digestlen);
+  ret = getrusage(RUSAGE_SELF, &ru1);
+  if (!PKCS5_PBKDF2_HMAC(pass, strlen(pass), salt, saltlen, rounds, digest,
+                         digestlen, buf)) {
+    explicit_bzero(buf, digestlen);
+    explicit_bzero(out, outlen);
+    putlog(LOG_MISC, "*", "PBKDF2 error: PKCS5_PBKDF2_HMAC(): %s.",
+           ERR_error_string(ERR_get_error(), NULL));
+    nfree(buf);
+    return NULL;
+  }
+  if (!ret && !getrusage(RUSAGE_SELF, &ru2)) {
+    debug4("pbkdf2 method %s rounds %i, user %.3fms sys %.3fms", digest_name,
+           rounds,
+           (double) (ru2.ru_utime.tv_usec - ru1.ru_utime.tv_usec) / 1000 +
+           (double) (ru2.ru_utime.tv_sec  - ru1.ru_utime.tv_sec ) * 1000,
+           (double) (ru2.ru_stime.tv_usec - ru1.ru_stime.tv_usec) / 1000 +
+           (double) (ru2.ru_stime.tv_sec  - ru1.ru_stime.tv_sec ) * 1000);
+  }
+  else {
+    debug1("PBKDF2 error: getrusage(): %s", strerror(errno));
+  }
+  if (b64_ntop_without_padding(buf, digestlen, out2, restlen) < 0) {
+    explicit_bzero(out, outlen);
+    putlog(LOG_MISC, "*", "PBKDF2 error: b64_ntop(hash).");
+    nfree(buf);
+    return NULL;
+  }
+  nfree(buf);
+  return out;
+}
+
+/* Return
+ *   hash = "$pbkdf2-<digest>$rounds=<rounds>$<salt>$<hash>" (PHC string format)
+ *     salt and hash = base64
+ *   NULL = error
+ */
+static char *pbkdf2_encrypt(const char *pass)
+{
+  unsigned char salt[PBKDF2_SALT_LEN];
+  static char *buf;
+
+  if (RAND_bytes(salt, sizeof salt) != 1) {
+    putlog(LOG_MISC, "*", "PBKDF2 error: RAND_bytes(): %s.",
+           ERR_error_string(ERR_get_error(), NULL));
+    return NULL;
+  }
+  if (!(buf = pbkdf2_hash(pass, pbkdf2_method, salt, sizeof salt,
+                          pbkdf2_rounds))) {
+    explicit_bzero(salt, sizeof salt);
+    return NULL;
+  }
+  explicit_bzero(salt, sizeof salt);
+  return buf;
+}
+
+/* Return
+ *   hash = "$pbkdf2-<digest>$rounds=<rounds>$<salt>$<hash>" (PHC string format)
+ *     salt and hash = base64
+ *     old encrypted = verify successful
+ *     new encrypted = verify successful, reenrypted with new parameters
+ *   NULL = verify failed
+ */
+static char *pbkdf2_verify(const char *pass, const char *encrypted)
+{
+  char method[sizeof pbkdf2_method],
+       b64salt[B64_NTOP_CALCULATE_SIZE(PBKDF2_SALT_LEN) + 1],
+       b64hash[B64_NTOP_CALCULATE_SIZE(256) + 1];
+  char format[40];
+  unsigned int rounds;
+  const EVP_MD *digest;
+  unsigned char salt[PBKDF2_SALT_LEN + 1];
+  int saltlen;
+  static char *buf;
+
+  if (snprintf(format, sizeof format,
```

Comment:
Just as a note: this might not be worth the confusion, splitting on $ is a lot clearer, and this can't parse arbitrary PHC strings, just the exact ones we wrote (no other parameters etc.). There's room for improvement here later then I guess. But fine for now, of course.

Reply:
So, how will you track this "wishlist item"? Are you gonna open an Issue for it?